apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PooledSession.cs

```diff
@@ -115,20 +140,236 @@ private PooledSession(SessionPool.ISessionPool pool, SessionName sessionName, By
+                        // Either way we need to create a transaction explicitly.
+                        if (inlinedTransactionExtractor is null || TransactionOptions.ModeCase == ModeOneofCase.PartitionedDml)
+                        {
+                            _transactionCreationTask = Task.Run(() => SetExplicitTransactionAsync(cancellationToken), cancellationToken);
+                            commandMaybeWithTransactionAsync = () => CommandWithTransactionAsync(cancellationToken);
+                        }
+                        // The calling command supports inlining.
+                        // We attempt inlining but if that fails, we create a transaction explicitly.
+                        else
+                        {
+                            // Create a task for executing the command which inlines transaction creation.
+                            // If this task succeeds we'll have both the transaction ID and the response from the command.
+                            // If this task fails we'll have to attempt to begin a transaction explicitly, which will give us a transaction ID,
+                            // and then we'll have to execute the command with that transaction.
+                            Task<TResponse> commandWithInliningTask = Task.Run(CommandWithInliningAsync, cancellationToken);
+
+                            // Now, create two tasks, one that is done when we have a transaction ID (via inlining or explicit),
+                            // and one that is done when the command is done (with inlining or explicit).
+
+                            // The transaction creation task is the combination of attempting inlining,
+                            // and if that fails, explicitly creating a transaction.
+                            _transactionCreationTask = Task.Run(async () =>
+                            {
+                                try
+                                {
+                                    await SetInlinedTransactionAsync(commandWithInliningTask).ConfigureAwait(false);
+                                }
+                                catch (Exception ex)
+                                {
+                                    Client.Settings.Logger.Warn($"Transaction creation via inlining failed. " +
+                                        $"Attempting to begin an explicit transaction. ",
+                                        ex);
+                                    await SetExplicitTransactionAsync(cancellationToken).ConfigureAwait(false);
+                                }
+                            }, cancellationToken);
+
+                            // The command execution task is the combination of attempting inlining,
+                            // and if that fails, executing the command with the explicitly created transaction.
+                            commandMaybeWithTransactionAsync = async () =>
```

Comment:
Perhaps move this to a local function? It's weird to review this but know that it's actually *not* called in the lock.